e2e/backend_service_update_test.go

```diff
@@ -0,0 +1,237 @@
+		for i, dst := range dests {
+			srvcs[i] = dst.ServiceName
+		}
+		if df := cmp.Diff(srvcs, rulesToBs[i], cmpopts.SortSlices(less)); df != "" {
+			t.Fatalf("Rule %d with action %+v  points to incorrect backend services, diff: %s , want nil", i, r.Action, df)
+		}
+	}
+}
+
+func checkBackendService(t *testing.T, ctx context.Context, cloud cloud.Cloud, bsID *cloud.ResourceID, bsPort int) {
+	t.Helper()
+	t.Log("---- Check BackendService ----")
+	gotBS, err := cloud.BackendServices().Get(ctx, bsID.Key)
+	if err != nil {
+		t.Fatalf("cloud.HealthChecks().Get(_, %s) = %v, want nil", bsID.Key, err)
+	}
+
+	if gotBS.Port != int64(bsPort) {
+		t.Errorf("BackendService port mismatch, got: %d want %d", gotBS.Port, bsPort)
+	}
+}
+func TestBackendServiceUpdate(t *testing.T) {
+	t.Parallel()
+
+	ctx := context.Background()
+	graphBuilder := rgraph.NewBuilder()
+
+	meshURL, meshKey := ensureMesh(ctx, t)
+	t.Cleanup(func() {
+		err := theCloud.Meshes().Delete(ctx, meshKey)
+		t.Logf("theCloud.Meshes().Delete(ctx, %s): %v", meshKey, err)
+	})
+
+	hc1ID, err := buildHealthCheck(graphBuilder, "hc1-test", 15)
+	if err != nil {
+		t.Fatalf("buildHealthCheck(_, hc1-test, 15) = (_, %v), want (_, nil)", err)
+	}
+	hc2ID, err := buildHealthCheck(graphBuilder, "hc2-test", 15)
+	if err != nil {
+		t.Fatalf("buildHealthCheck(_, hc2-test, 15) = (_, %v), want (_, nil)", err)
+	}
+
+	bs1Name := "bs1-e2e"
+	bs2Name := "bs2-e2e"
+
+	bs1ID, err := buildBackendServiceWithLBScheme(graphBuilder, bs1Name, hc1ID, "INTERNAL_SELF_MANAGED")
+	if err != nil {
+		t.Fatalf("buildBackendServiceWithLBScheme(_, %s, _) = %v, want nil", resourceName(bs1Name), err)
+	}
+
+	bs2ID, err := buildBackendServiceWithLBScheme(graphBuilder, bs2Name, hc2ID, "INTERNAL_SELF_MANAGED")
+	if err != nil {
+		t.Fatalf("buildBackendServiceWithLBScheme(_, %s, _) = %v, want nil", resourceName(bs2Name), err)
+	}
+
+	rules := []*networkservices.TcpRouteRouteRule{
+		{
+			Action: &networkservices.TcpRouteRouteAction{
+				Destinations: []*networkservices.TcpRouteRouteDestination{
+					{
+						ServiceName: resourceSelfLink(bs1ID),
+						Weight:      10,
+					},
+				},
+			},
+			Matches: []*networkservices.TcpRouteRouteMatch{
+				{
+					Address: routeCIDR,
+					Port:    "80",
+				},
+			},
+		},
+		{
+			Action: &networkservices.TcpRouteRouteAction{
+				Destinations: []*networkservices.TcpRouteRouteDestination{
+					{
+						ServiceName: resourceSelfLink(bs2ID),
+						Weight:      10,
+					},
+				},
+			},
+			Matches: []*networkservices.TcpRouteRouteMatch{
+				{
+					Address: routeCIRD2,
+					Port:    "80",
+				},
+			},
+		},
+	}
+	serviceURL := fmt.Sprintf("https://compute.googleapis.com/v1/projects/%s/global/backendServices/%s", testFlags.project, resourceName(bs1Name))
+	tcpr, err := buildTCPRoute(graphBuilder, "test-route", meshURL, rules, bs1ID)
+	t.Logf("tcpr = %s", pretty.Sprint(tcpr))
+	if err != nil {
+		t.Fatalf("buildTcpRoute(_, route1, %s) = %v, want nil", serviceURL, err)
```

Comment:
please change params for buildTCPRoute. you don't have to log all params if they are complex.